src/lsm/compaction.zig

```diff
@@ -42,48 +42,313 @@ const tracer = @import("../tracer.zig");
 const schema = @import("schema.zig");
-const TableDataIteratorType = @import("table_data_iterator.zig").TableDataIteratorType;
-const LevelTableValueBlockIteratorType =
-    @import("level_data_iterator.zig").LevelTableValueBlockIteratorType;
+
+/// The upper-bound count of input tables to a single tree's compaction.
+///
+/// - +1 from level A.
+/// - +lsm_growth_factor from level B. The A-input table cannot overlap with an extra B-input table
+///   because input table selection is least-overlap. If the input table overlaps on one or both
+///   edges, there must be another table with less overlap to select.
+pub const compaction_tables_input_max = 1 + constants.lsm_growth_factor;
+
+/// The upper-bound count of output tables from a single tree's compaction.
+/// In the "worst" case, no keys are overwritten/merged, and no tombstones are dropped.
+pub const compaction_tables_output_max = compaction_tables_input_max;
+
+const half_bar_beat_count = @divExact(constants.lsm_batch_multiple, 2);
+
+/// Information used when scheduling compactions. Kept unspecalized to make the forest
+/// code easier.
+pub const CompactionInfo = struct {
+    /// How many values, across all tables, need to be processed.
+    compaction_tables_value_count: usize,
+
+    // Keys are integers in TigerBeetle, with a maximum size of u256. Store these
+    // here, instead of Key, to keep this unspecalized.
+    target_key_min: u256,
+    target_key_max: u256,
+
+    /// Are we doing a move_table? In which case, certain things like grid reservation
+    /// must be skipped by the caller.
+    move_table: bool,
+
+    level_b: u8,
+};
+
+pub const Exhausted = struct { bar: bool, beat: bool };
+const BlipCallback = *const fn (*anyopaque, ?Exhausted) void;
+pub const BlipStage = enum { read, merge, write, drained };
+
+// The following types need to specalize on Grid, but are used both by CompactionType and the
+// forest.
+pub fn CompactionHelperType(comptime Grid: type) type {
+    return struct {
+        pub const CompactionBlocks = struct {
+            /// Index blocks are global, and shared between blips. The index reads happen
+            /// as a mini-stage before reads kick off.
+            source_index_blocks: []CompactionBlock,
+
+            /// For each source level, we have a buffer of CompactionBlocks.
+            source_value_blocks: [2]BlockFIFO,
+
+            /// We only have one buffer of output CompactionBlocks.
+            target_value_blocks: BlockFIFO,
+        };
+
+        pub const CompactionBlock = struct {
+            block: BlockPtr,
+
+            // CompactionBlocks are stored in buffers where we need a pointer to get back to our
+            // parent.
+            target: ?*anyopaque = null,
+
+            // TODO: This could be a union to save a bit of memory and add a bit of safety.
+            read: Grid.Read = undefined,
+            write: Grid.Write = undefined,
+
+            next: ?*CompactionBlock = null,
+        };
```

Comment:
Consider adding

```zig
stage: enum { free, pending, ready, ioing }
```

field to `CompactionBlock`  and asserting that the block is at a correct stage on use.